File: ios/NetworkManager.swift
// Copyright © 2023 Middleware. Licensed under the Apache License, Version 2.0
#if os(iOS) || targetEnvironment(macCatalyst) || os(tvOS)
import UIKit
import SWCompression

let IMAGES_URL = "/v1/rum"

class NetworkManager: NSObject {
    public var baseUrl: String? = nil
    
    public var sessionId: String? = nil
    private var token: String? = nil
    
    init(target: String, token: String) {
        self.baseUrl = target
        self.token = token
    }

    private func createRequest(method: String, path: String) -> URLRequest {
        let url = URL(string: self.baseUrl!+path)!
        var request = URLRequest(url: url)
        request.httpMethod = method
        return request
    }

    private func callAPI(request: URLRequest,
                 onSuccess: @escaping (Data) -> Void,
                 onError: @escaping (Error?) -> Void) {
    
        let task = URLSession.shared.dataTask(with: request) { (data, response, error) in
            DebugUtils.log(">>>\(request.httpMethod ?? ""):\(request.url?.absoluteString ?? "")\n<<<\(String(data: data ?? Data(), encoding: .utf8) ?? "")")
            
            DispatchQueue.main.async {
                guard let data = data,
                      let httpResponse = response as? HTTPURLResponse,
                      (200...299).contains(httpResponse.statusCode) else {
                    DebugUtils.error(">>>>>> Error in call \(request.url?.absoluteString ?? "") : \(error?.localizedDescription ?? "N/A")")
                    onError(error)
                    return
                }
                onSuccess(data)
            }
        }
        task.resume()
    }
    
    
    func sendImages(sessionId: String, images: Data, name: String, completion: @escaping (Bool) -> Void) {
        var request = createRequest(method: "POST", path: IMAGES_URL)
        guard let token = self.token else {
            completion(false)
            return
        }
        request.setValue("\(token)", forHTTPHeaderField: "Authorization")
        let boundary = "Boundary-\(NSUUID().uuidString)"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        let parameters = ["sessionId": sessionId]
        for (key, value) in parameters {
            body.appendString("--\(boundary)\r\n")
            body.appendString("Content-Disposition: form-data; name=\"\(key)\"\r\n\r\n")
            body.appendString("\(value)\r\n")
        }

        body.appendString("--\(boundary)\r\n")
        body.appendString("Content-Disposition: form-data; name=\"batch\"; filename=\"\(name)\"\r\n")
        body.appendString("Content-Type: gzip\r\n\r\n")
        body.append(images)
        body.appendString("\r\n")

        body.appendString("--\(boundary)--\r\n")
        DebugUtils.log(">>>>>> sending \(body.count) bytes")
        request.httpBody = body

        callAPI(request: request) { (data) in
            completion(true)
        } onError: { _ in
            completion(false)
        }
    }
}
#endif

File: android/sdk/src/main/java/io/middleware/android/sdk/core/replay/v2/NetworkManager.java
package io.middleware.android.sdk.core.replay.v2;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okio.BufferedSink;
import okio.GzipSink;
import okio.Okio;

public class NetworkManager {
    private static final String IMAGES_URL = "/v1/rum";

    private final String baseUrl;
    private final String token;
    private final Handler handler = new Handler(Looper.getMainLooper());

    public NetworkManager(String baseUrl, String token) {
        this.baseUrl = baseUrl;
        this.token = token;
    }

    private Request createRequest(final String path) {
        final String url = baseUrl + path;
        return new Request.Builder()
                .url(url)
                .build();
    }

    private void callAPI(Request request, final NetworkCallback callback) {
        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(new GzipRequestInterceptor())
                .build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                final String responseData = response.body() != null ? response.body().string() : "";
                final int statusCode = response.code();
                handler.post(() -> {
                    if (statusCode >= 200 && statusCode < 300) {
                        callback.onSuccess(responseData);
                    } else {
                        callback.onError(new IOException("Error in call: " + statusCode));
                    }
                });
            }

            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                handler.post(() -> callback.onError(e));
            }
        });
    }

    public void sendImages(String sessionId, byte[] images, String name, final NetworkCallback callback) {
        Request request = createRequest(IMAGES_URL);
        if (token == null) {
            callback.onError(new IOException("Token is null"));
            return;
        }


        String boundary = "Boundary-" + java.util.UUID.randomUUID().toString();
        MultipartBody.Builder requestBodyBuilder = new MultipartBody.Builder(boundary)
                .setType(MultipartBody.FORM)
                .addFormDataPart("sessionId", sessionId)
                .addFormDataPart("batch", name,
                        RequestBody.create(images, MediaType.parse("application/gzip")));

        RequestBody requestBody = requestBodyBuilder.build();
        request = request.newBuilder()
                .post(requestBody)
                .header("Authorization", token)
                .build();

        callAPI(request, new NetworkCallback() {
            @Override
            public void onSuccess(String response) {
                callback.onSuccess(response);
            }

            @Override
            public void onError(Exception e) {
                callback.onError(e);
            }
        });
    }

    static class GzipRequestInterceptor implements Interceptor {
        @Override
        public Response intercept(Chain chain) throws IOException {
            Request originalRequest = chain.request();
            if (originalRequest.body() == null || originalRequest.header("Content-Encoding") != null) {
                return chain.proceed(originalRequest);
            }

            Request compressedRequest = originalRequest.newBuilder()
                    .header("Content-Encoding", "gzip")
                    .method(originalRequest.method(), gzip(originalRequest.body()))
                    .build();
            return chain.proceed(compressedRequest);
        }

        private RequestBody gzip(final RequestBody body) {
            return new RequestBody() {
                @Override
                public MediaType contentType() {
                    return body.contentType();
                }

                @Override
                public long contentLength() {
                    return -1; // We don't know the compressed length in advance!
                }

                @Override
                public void writeTo(BufferedSink sink) throws IOException {
                    BufferedSink gzipSink = Okio.buffer(new GzipSink(sink));
                    body.writeTo(gzipSink);
                    gzipSink.close();
                }
            };
        }
    }
}
